control_panel_api/authentication.py

```diff
@@ -1,37 +1,69 @@
+import logging
+
+import requests
 from django.conf import settings
-import jwt
-from jwt.exceptions import InvalidTokenError
+from jose import jwt
+from jose.exceptions import JWTError
+from requests.exceptions import RequestException
 from rest_framework.authentication import (
     BaseAuthentication,
-    get_authorization_header
+    get_authorization_header,
 )
 from rest_framework.exceptions import AuthenticationFailed
 
 from control_panel_api.models import User
 
+logger = logging.getLogger(__name__)
 
-class Auth0JWTAuthentication(BaseAuthentication):
 
-    def authenticate(self, request):
+def get_jwks():
+    response = requests.get(settings.OIDC_WELL_KNOWN_URL)
+    return response.json()
+
+
+def get_key(jwks, unverified_header):
+    for jwk in jwks.get('keys'):
+        if jwk['kid'] == unverified_header['kid']:
+            return jwk
 
+    return None
+
+
+class Auth0JWTAuthentication(BaseAuthentication):
+    def authenticate(self, request):
         try:
             prefix, token = get_authorization_header(request).split()
-
         except (ValueError, TypeError):
             return None
 
         if prefix != 'JWT'.encode():
             return None
 
+        try:
+            unverified_header = jwt.get_unverified_header(token)
+        except JWTError as e:
+            raise AuthenticationFailed('Error decoding JWT header') from e
+
+        try:
```

Comment:
I'd move the handling of this exception into `get_jwks()` to streamline this method.

Reply:
I agree with this one - partially. `get_jwks` should maybe catch and log (esp. to Sentry) the `RequestException`, but either
1. raise a new more specific exception (eg: `JwksFetchError`), or
2. return an empty dict to pass to `get_key` and let the authentication fail because the key could not be matched.